resources/type-system/upper-lower-bounds.md

```diff
@@ -102,22 +105,24 @@ We define the upper bound of two types T1 and T2 to be **UP**(`T1`,`T2`) as foll
 - **UP**(`T1`, `T2`) where **OBJECT**(`T1`) =
   - `T1` if `T2` is non-nullable
   - `T1*` if `Null <: T2` (that is, `T2` is legacy)
-  - `T1?` otherwise
+  - `T1?` otherwise.
 
 - **UP**(`T1`, `T2`) where **OBJECT**(`T2`) =
   - `T2` if `T1` is non-nullable
   - `T2*` if `Null <: T1` (that is, `T1` is legacy)
-  - `T2?` otherwise
-
-- **UP**(`T1*`, `T2*`) = `S*` where `S` is **UP**(`T1`, `T2`)
-- **UP**(`T1*`, `T2?`) = `S?` where `S` is **UP**(`T1`, `T2`)
-- **UP**(`T1?`, `T2*`) = `S?` where `S` is **UP**(`T1`, `T2`)
-- **UP**(`T1*`, `T2`) = `S*` where `S` is **UP**(`T1`, `T2`)
-- **UP**(`T1`, `T2*`) = `S*` where `S` is **UP**(`T1`, `T2`)
-
-- **UP**(`T1?`, `T2?`) = `S?` where `S` is **UP**(`T1`, `T2`)
-- **UP**(`T1?`, `T2`) = `S?` where `S` is **UP**(`T1`, `T2`)
-- **UP**(`T1`, `T2?`) = `S?` where `S` is **UP**(`T1`, `T2`)
+  - `T2?` otherwise.
+
+- **UP**(`T1*`, `T2*`) = `S*` where `S` is **UP**(`T1`, `T2`).
+- **UP**(`T1*`, `T2?`) = `S?` where `S` is **UP**(`T1`, `T2`).
+- **UP**(`T1?`, `T2*`) = `S?` where `S` is **UP**(`T1`, `T2`).
+- **UP**(`T1*`, `T2`) = `S*` where `S` is **UP**(`T1`, `T2`).
+- **UP**(`T1`, `T2*`) = `S*` where `S` is **UP**(`T1`, `T2`).
+
+- **UP**(`T1?`, `T2?`) = `S?` where `S` is **UP**(`T1`, `T2`).
+- **UP**(`T1?`, `T2`) = `S?` where `S` is **UP**(`T1`, `T3`)
+  where `T3` is **ERASEINTERSECTION**(`T2`).
```

Comment:
Could we make this

... where `S` is **ERASEINTERSECTION**(**UP**(`T1`,`T2`)).

Then **UP**(`int?`, `X&int`) could possibly be `int?`, which or can't when we erase too eagerly.

I don't know if that'll work today, because I haven't been able to memorize the rules, but even if not. maybe it will in the future (say of we improve **UP**(`X&T1`, `X&T3`) at some point).

And we won't get a different result for **UP**(`T1`,`T2?`) than for **UP**(`T1?`,`T2?`).
The only difference is that currently we know that the `T`of `T?` cannot be an intersection type, and the **UP** of two inon-ntersection types cannot be an intersection type, so we don't need to mention **ERASEINTERSECTION** in the  **UP**(`T1?`,`T2?`) case. But maybe we should, just to be consistent and future proof. "Always erase before embedding" into a composite type, even in the places where it's _currently_ never needed.

That is, generally, erase an intersection just before adding the `?`, but no sooner.
That should still be safe, and we'll get the benefit of any improvements we do wrt. Intersection types. And if we write it everywhere, we won't forget a case later if we start introducing intersection types, somehow.